### PR TITLE
Create dangling ingress check 

### PR DESCRIPTION
**Pull request by charlesoconor**
Creating a check for dangling ingresses. I think it will fit in well with the other rules that already exist in this tool. Have been seeing these pop up from time to time in the projects I've been working on and would love to be able to catch them at lint time. 

I punted on dealing with resources that ingress can point to an only check for services, but could add that logic in if need be. My assumption was this would cover almost all of the use cases. 

Closes https://github.com/stackrox/kube-linter/issues/387

### CODE REVIEW

**Comment by janisz** [comment]
Could you add e2e test?

**Comment by charlesoconor** [comment]
Yes we can. Adding a e2e test

File: pkg/templates/danglingingress/template.go [hunk 1]
```diff
@@ -0,0 +1,112 @@
...
+					service, ok := obj.K8sObject.(*v1.Service)
+					if !ok {
+						continue
+					}
+
+					serviceName := service.ObjectMeta.Name
+					if _, ok := selectors[serviceName]; ok {
+						delete(selectors, serviceName)
+						if len(selectors) == 0 {
+							// Found the all!
```
Comment (dhaus67) [inline pkg/templates/danglingingress/template.go:94]: super nit:
```suggestion
							// Found them all!
```

File: pkg/templates/danglingingress/template.go [hunk 2]
```diff
@@ -0,0 +1,112 @@
...
+	service := b.Service
+	if service == nil {
+		return ""
+	}
+
+	return service.Name
+}
+
+func getSelectorsFromIngress(ingress *networkingV1.Ingress) map[string]bool {
+	selectors := map[string]bool{}
```
Comment (dhaus67) [inline pkg/templates/danglingingress/template.go:31]: I'd favor `map[string]struct{}` and setting `selectors[s] = struct{}{}` subsequently:
```suggestion
	selectors := map[string]struct{}
```

**Review by dhaus67 — COMMENTED** [review]
Thanks for the contrib + adding the E2E tests so fast!

I would have one question regarding the definition of dangling ingress resource:
Currently, we define the ingress rule as dangling if the service name does not match. 

However, for the ingress to work correctly, the service's port name / number must match as well. This can easily be misconfigured.

I saw this multiple times in the past be the culprit of the ingress not working correctly, so I do see the opportunity here to improve this as well with the new check during linting.

I'd suggest we add this to the dangling check as well. Then, we will check that both the service name match as well as defined port (either via its name / number).

File: pkg/lintcontext/mocks/ingress.go [hunk 3]
```diff
@@ -0,0 +1,29 @@
...
+			Kind:       objectkinds.Ingress,
+			APIVersion: objectkinds.GetIngressAPIVersion(),
+		},
+		ObjectMeta: metaV1.ObjectMeta{Name: name},
+		Spec:       networkingV1.IngressSpec{},
+	}
+}
+
+// ModifyIngess modifies a given networkpolicy in the context via the passed function.
+func (l *MockLintContext) ModifyIngess(t *testing.T, name string, f func(ingress *networkingV1.Ingress)) {
```
[the remark this answers is not in the document]
Comment (charlesoconor) [reply pkg/lintcontext/mocks/ingress.go]: :+1:

File: pkg/templates/danglingingress/template.go (hunk 1 above)
Comment (charlesoconor) [reply pkg/templates/danglingingress/template.go]: :+1:

File: pkg/templates/danglingingress/template.go (hunk 2 above)
Comment (charlesoconor) [reply pkg/templates/danglingingress/template.go]: :+1:

File: pkg/templates/danglingingress/template.go [hunk 4]
```diff
@@ -0,0 +1,112 @@
...
+	return selectors
+}
+
+func init() {
+	templates.Register(check.Template{
+		HumanName:   "Dangling Ingress",
+		Key:         templateKey,
+		Description: "Flag ingress which do not match any service",
+		SupportedObjectKinds: config.ObjectKindsDesc{
+			ObjectKinds: []string{objectkinds.Service},
```
[the remark this answers is not in the document]
Comment (charlesoconor) [reply pkg/templates/danglingingress/template.go]: yes it should be

**Comment by charlesoconor** [comment]
I think it's a great idea to check the ports as well. Update the PR for that.

File: docs/generated/checks.md [hunk 5]
```diff
@@ -67,6 +67,15 @@ verbs:
 **Remediation**: Confirm that your HorizontalPodAutoscaler's scaleTargetRef correctly matches one of your deployments.
 
 **Template**: [dangling-horizontalpodautoscaler](generated/templates.md#dangling-horizontalpodautoscalers)
+## dangling-ingress
+
+**Enabled by default**: No
+
+**Description**: Indicates when ingress do not have any associated services.
+
+**Remediation**: Confirm that your ingress's backend correctly matches the name on one of your services.
```
Comment (dhaus67) [inline docs/generated/checks.md:76]:
```suggestion
**Remediation**: Confirm that your ingress's backend correctly matches the name and port on one of your services.
```

File: pkg/builtinchecks/yamls/dangling-ingress.yaml [hunk 6]
```diff
@@ -0,0 +1,7 @@
+name: "dangling-ingress"
+description: "Indicates when ingress do not have any associated services."
+remediation: "Confirm that your ingress's backend correctly matches the name on one of your services."
```
Comment (dhaus67) [inline pkg/builtinchecks/yamls/dangling-ingress.yaml:3]:
```suggestion
remediation: "Confirm that your ingress's backend correctly matches the name and port on one of your services."
```

**Review by dhaus67 — APPROVED** [review]
Only two minor doc comments, otherwise LGTM!

Thanks for making the changes quickly!

File: docs/generated/checks.md (hunk 5 above)
Comment (charlesoconor) [reply docs/generated/checks.md:76]: :+1:

File: pkg/builtinchecks/yamls/dangling-ingress.yaml (hunk 6 above)
Comment (charlesoconor) [reply pkg/builtinchecks/yamls/dangling-ingress.yaml]: :+1:

**Comment by charlesoconor** [comment]
Thanks for the approval! Excited to be able to use this check I've been running into issues around this a lot lately.

**Pull request merged**